### PR TITLE
Monitoring infrastructure

### PR DESCRIPTION
**Pull request by bosilca**
Bring the curated point-to-point monitoring infrastructure to devel.

### CODE REVIEW

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by jsquyres** [comment]
bot:retest

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by jsquyres** [comment]
@bosilca The distcheck test is failing with:

```
make[3]: Entering directory `/var/lib/jenkins/workspace/ompi_master_pr_distcheck/ompi/mca/pml/monitoring'
make[3]: *** No rule to make target `pml_monitoring_comm.h', needed by `distdir'.  Stop.
```

Can you fix?

**Comment by jsquyres** [comment]
@bosilca Looks like a rebase went wrong on this PR.  Can you fix?

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by jsquyres** [comment]
@bosilca Want to look at this in Bordeaux next week?  :-)

**Comment by bosilca** [comment]
No, I just didn't bother fixing it as I am still working on improving the monitoring system.

  George.

**Comment by hppritcha** [comment]
I'd suggest starting over in your workarea with a fresh branch of master, then cherry-pick the commits that actually go in to this PR in to the new branch.

This looks like a merge of master was done in to the branch rather than a rebase.

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by bosilca** [comment]
@hjelmn please review, especially the commits I got from your branch (5b3a1ee) and the modification to PVAR I did.

**Comment by bosilca** [comment]
@hjelmn ping.

**Comment by hjelmn** [comment]
Will try to get through this today.

File: ompi/mca/pml/monitoring/pml_monitoring.c [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+                my_rank = i;
+            key = *((uint64_t*)&(procs[i]->super.proc_name));
+            /* store the rank (in COMM_WORLD) of the process
+               with its name (a uniq opal ID) as key  in the hash table*/
+            opal_hash_table_set_value_uint64(translation_ht,
+                                             key,
+                                             (void*)(uintptr_t)i);
+        }
+    }
+    return pml_selected_module.pml_add_procs(procs, nprocs);
```
Comment (hjelmn) [inline ompi/mca/pml/monitoring/pml_monitoring.c]: Keep in mind that when dynamic add_procs is in use the above code will only add the local ranks to translation_ht. It looks like you don't need the procs, just the opal_process_name_t's. You can generate those yourself as is done at ompi/proc/proc.c:337 .

Comment (hjelmn) [inline ompi/mca/pml/monitoring/pml_monitoring.c]: Alternatively you can set the MCA_PML_BASE_FLAG_REQUIRE_WORLD flag in the pml_flags field to disable dynamic add_procs when using the monitoring component.

Comment (bosilca) [inline ompi/mca/pml/monitoring/pml_monitoring.c]: Thanks for catching this. We don't need to populate the hash_table during add_proc, we can do it lazily when we need a particular proc.

Comment (bosilca) [inline ompi/mca/pml/monitoring/pml_monitoring.c]: @hjelmn I rewrote the add_procs to work with the dynamic proc management. Please take a look.

**Comment by bosilca** [comment]
I plan to merge this in on Thursday morning EST. Any comments before that will be welcome.

**Comment by hjelmn** [comment]
I see one more things that could be improved. You should probably set the pml_flags for modules in the monitoring component based on the underlying pml. That way if the underlying pml set the flag indicating that it does not support dynamic add_procs the flag preserved.

**Comment by bosilca** [comment]
@hjelmn thanks for catching up the missing flags. I also restored the context id, tag and progress.

**Pull request merged**